files/initramfs-tools/union-mount.j2

```diff
@@ -39,10 +39,49 @@ set_tmpfs_log_partition_size()
   [ $maxsize -le $varlogsize ] && varlogsize=$maxsize
 }
 
+whitelist_rw_folder()
+{
+  image_dir=$1
+  whitelist_file=${rootmnt}/host/$image_dir/whitelist
+ 
+  # Return if the whitelist file does not exist
+  if ! test -f "${whitelist_file}"; then
+    return
+  fi
+ 
+  # Return if the secure_boot_enable option is not set
+  if cat /proc/cmdline | grep -v -q "secure_boot_enable=true"; then
+    return
+  fi
+ 
+  whitelist_log=${rootmnt}/host/$image_dir/whitelist.log
+  rw_dir=${rootmnt}/host/$image_dir/rw
+  whitelist=$(cat ${rootmnt}/host/$image_dir/whitelist | grep -v "^\s*#" | awk '{$1=$1};1')
+  set -o noglob
+  find ${rw_dir} -type f |
+  while IFS= read -r file; do
+    found="false"
+    for line in $whitelist; do
+      pattern="^${rw_dir}/${line}\$"
+      if echo "$file" | grep -q "$pattern"; then
+        found="true"
+        break
+      fi
+    done
+    if [ $found = "false" ]; then
+      echo $file >> ${whitelist_log}
+      rm -f $file
+    fi
+  done
```

Comment:
I am worried about the performance if `rw` contains many files and deep directories. Since whitelist is short, can we `mv` folder to a temp one, and copied back one by one.

Not sure the validation of this solution, is there any metadata / hidden files in rw for overlay?

Reply:
The find command can get hidden files, what is the metadata?

Reply:
Yes, the performance is not good enough, but it is not relative to the depth of the directories, it is the grep.

Reply:
The performance issue should be fixed. Tested on the device, reduce from 100 seconds to 1 seconds for 300 files. And checked the changed files on product, it only has about 700 changed files in max.